source/base/Shmoose.cpp

```diff
@@ -353,3 +375,23 @@ void Shmoose::removeRoom(QString const &roomJid)
 {
     mucManager_->removeRoom(roomJid);
 }
+
+void Shmoose::attachmentUploadFailed()
+{
+    persistence_->markMessageAsSendFailed(notSentMsgId_);
+    notSentMsgId_ = "";
+}
+
+void Shmoose::saveAttachment(const QString& msg)
+{
+    //TODO Error management + Destination selection
+    QFile::copy(getLocalFileForUrl(msg), QStandardPaths::locate(QStandardPaths::DownloadLocation, "", QStandardPaths::LocateDirectory)  +
+                    QDir::separator() + CryptoHelper::getHashOfString(QUrl(msg).toString(), true));
+}
+
+void Shmoose::fileUploaded(QString const&toJid, QString const&message, QString const&type)
+{
+    persistence_->removeMessage(notSentMsgId_);
```

Comment:
Even it should not happen, I have seen msgId's which are doubled. So it is wise to also include the toJid as a parameter to the removeMessage methode and foreward this also to the MessageControler::removeMessage methode to do the actual sql delete query.